File: node-version/src/logging/FileLogger.js
import fs from 'fs/promises';
import path from 'path';

class FileLogger {

    constructor(logDirectory) {

        this.logDirectory = logDirectory;

        const timestamp = this.generateTimestamp();

        this.logFilePath = path.join(
            logDirectory,
            `diagnostic_${timestamp}.log`
        );
    }

    async initialize() {

        await fs.mkdir(
            this.logDirectory,
            { recursive: true }
        );
    }

    generateTimestamp() {

        const now = new Date();

        const pad = value =>
            value.toString().padStart(2, '0');

        return (
            now.getFullYear() +
            pad(now.getMonth() + 1) +
            pad(now.getDate()) +
            '_' +
            pad(now.getHours()) +
            pad(now.getMinutes()) +
            pad(now.getSeconds())
        );
    }

    async logNetworkProbe(result) {

        await this.writeLine(
            `[NETWORK] ${result.toString()}`
        );
    }

    async logDatabaseProbe(result) {

        await this.writeLine(
            `[DATABASE] ${result.toString()}`
        );
    }

    async logAnomaly(anomaly) {

        await this.writeLine(
            `[ANOMALY] ${anomaly.toString()}`
        );
    }

    async logSummary(summary) {

        await this.writeLine('');

        await this.writeLine(
            '===== FINAL SUMMARY ====='
        );

        await this.writeLine(summary);
    }

    async writeLine(message) {

        try {

            await fs.appendFile(
                this.logFilePath,
                `${message}\n`
            );

        } catch (error) {

            console.error(
                `Failed to write log: ${error.message}`
            );
        }
    }

    getLogFilePath() {
        return this.logFilePath;
    }
}

export default FileLogger;

File: src/main/java/org/example/logging/FileLogger.java
package org.example.logging;

import org.example.model.AnomalyEvent;
import org.example.model.DbResult;
import org.example.model.ProbeResult;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class FileLogger {

    private final String logFilePath;

    public FileLogger(String logDirectory) {

        File directory = new File(logDirectory);

        if (!directory.exists()) {
            directory.mkdirs();
        }

        String timestamp = LocalDateTime.now()
                .format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));

        this.logFilePath =
                logDirectory +
                        File.separator +
                        "diagnostic_" +
                        timestamp +
                        ".log";
    }

    public void logNetworkProbe(ProbeResult result) {

        writeLine(
                "[NETWORK] " +
                        result.toString()
        );
    }

    public void logDatabaseProbe(DbResult result) {

        writeLine(
                "[DATABASE] " +
                        result.toString()
        );
    }

    public void logAnomaly(AnomalyEvent anomalyEvent) {

        writeLine(
                "[ANOMALY] " +
                        anomalyEvent.toString()
        );
    }

    public void logSummary(String summary) {

        writeLine("");
        writeLine("===== FINAL SUMMARY =====");
        writeLine(summary);
    }

    private void writeLine(String message) {

        try (BufferedWriter writer =
                     new BufferedWriter(
                             new FileWriter(logFilePath, true))) {

            writer.write(message);
            writer.newLine();

        } catch (IOException e) {

            System.err.println(
                    "Failed to write log: "
                            + e.getMessage()
            );
        }
    }

    public String getLogFilePath() {
        return logFilePath;
    }
}